specs/driver_spec.rb

```diff
@@ -0,0 +1,225 @@
+      last_driver.vin.must_equal 'XF9Z0ST7X18WD41HT'
+    end
+  end
+
+                        ###########################
+                        ## instance method specs ##
+                        ###########################
+
+  # wrote this loop to sort trips by driver - this returns an array of arrays
+  # each nested array contains trips for same driver - can use to find driver
+  # with one trip and the most trips
+  def sort_trips_by_driver(trips, number_of_drivers)
+    trips_sorted_by_driver_id = Array.new(number_of_drivers) { |i| i = [] }
+    trips.each do |trip|
+      number_of_drivers.times do |i|
+        trips_sorted_by_driver_id[i] << trip if trip.driver_id == i
+      end
+    end
+    trips_sorted_by_driver_id
+  end
+
+  before do
+    # hand calculated driver_id 21's average: 30.0/11 = 2.73
+    @driver_id = 21
+    trips = RideShare::Trip.all
+    number_of_drivers = RideShare::Driver.all.length
+    trips_sorted_by_driver_id = sort_trips_by_driver(trips, number_of_drivers)
+    # drivers 27 and 37 have only 1 trip - driver 27's one trip rating is 1
+    @driver_with_one_trip = trips_sorted_by_driver_id.min_by { |array| array.length }
+    # drivers: 54 has most trips (12 of them)
+    @driver_with_most_trips = trips_sorted_by_driver_id.max_by { |array| array.length }
+  end
+
+  let(:drivers) {RideShare::Driver.all}
+  let(:driver_know_avg) { RideShare::Driver.find(@driver_id) }
+  let(:driver_no_trips) { RideShare::Driver.find(100) }
+  let(:driver_1_trip) { RideShare::Driver.find(@driver_with_one_trip[0].driver_id)}
+  let(:drivers_with_most_trips) {RideShare::driver.find(@drivers_with_most_trips[0].driver_id)}
+
+  describe "Driver#get_trips" do
+
+    it "returns an array of trip instances" do
+      driver_know_avg.get_trips.must_be_instance_of Array
+      driver_know_avg.get_trips.each { |trip| trip.must_be_instance_of RideShare::Trip }
+    end
+
+    it "each trip instance has the same driver id" do
+      driver_know_avg.get_trips.each { |trip| trip.driver_id.must_equal @driver_id }
+    end
+  end
+
+  describe "Driver#calculate_average_rating" do
+
+    # used sample here to get a sample of the trips and makes sure they are
+    # returning an average in the expected range of 1-5
+    # could test all drivers for this data, but if data was significantly larger
+    # it might take too much time to test them all
+    it "returns a float between 1 and 5" do
+      drivers.sample(25).each do |driver|
+        return nil if driver.calculate_average_rating.nil?
+        driver.calculate_average_rating.must_be :>=, 1
+        driver.calculate_average_rating.must_be :<=, 5
+        driver.calculate_average_rating.must_be_instance_of Float
+      end
+    end
+
+    it "calculates the correct average" do
+      # this is the average for driver_id 21 calculated by hand
+      driver_know_avg.calculate_average_rating.must_equal 2.73
+    end
+
+    it "returns nil if there are no trips for driver" do
```

Comment:
Good test case! Edge cases like this, asking what happens when there *isn't* any data, can be easy to miss.